destination_middleware.go

```diff
@@ -498,3 +505,239 @@ func (d *destinationWithRecordFormat) Write(ctx context.Context, recs []opencdc.
+		errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		// log warning once, to avoid spamming the logs
+		d.keyWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record does not have an attached schema for the key, consider disabling the destination schema key decoding using "%s: false"`, configDestinationWithSchemaExtractionKeyEnabled)
+		})
+		return nil
+	case errSubject != nil && !errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema subject from metadata: %w", errSubject)
+	case errVersion != nil && !errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema version from metadata: %w", errVersion)
+	}
+
+	if rec.Key != nil {
+		decodedKey, err := d.decode(ctx, rec.Key, subject, version)
+		if err != nil {
+			return fmt.Errorf("failed to decode key: %w", err)
+		}
+		rec.Key = decodedKey
+	}
+
+	return nil
+}
+
+func (d *destinationWithSchemaExtraction) decodePayload(ctx context.Context, rec *opencdc.Record) error {
+	subject, errSubject := rec.Metadata.GetPayloadSchemaSubject()
+	version, errVersion := rec.Metadata.GetPayloadSchemaVersion()
+	switch {
+	case errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound) &&
+		errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		// log warning once, to avoid spamming the logs
+		d.payloadWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record does not have an attached schema for the payload, consider disabling the destination schema payload decoding using "%s: false"`, configDestinationWithSchemaExtractionPayloadEnabled)
+		})
+		return nil
+	case errSubject != nil && !errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get payload schema subject from metadata: %w", errSubject)
+	case errVersion != nil && !errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get payload schema version from metadata: %w", errVersion)
+	}
+
+	if rec.Payload.Before != nil {
+		decodedPayloadBefore, err := d.decode(ctx, rec.Payload.Before, subject, version)
+		if err != nil {
+			return fmt.Errorf("failed to decode payload.before: %w", err)
+		}
+		rec.Payload.Before = decodedPayloadBefore
+	}
+	if rec.Payload.After != nil {
+		decodedPayloadAfter, err := d.decode(ctx, rec.Payload.After, subject, version)
+		if err != nil {
+			return fmt.Errorf("failed to decode payload.after: %w", err)
+		}
+		rec.Payload.After = decodedPayloadAfter
+	}
+
+	return nil
+}
+
+func (d *destinationWithSchemaExtraction) decode(ctx context.Context, data opencdc.Data, subject string, version int) (opencdc.StructuredData, error) {
+	switch data := data.(type) {
+	case opencdc.StructuredData:
+		return data, nil // already decoded
+	case opencdc.RawData: // let's decode it
+	default:
+		return nil, fmt.Errorf("unexpected data type %T", data)
+	}
+
+	// check if we have the schema in the cache
+	if _, ok := d.subjectVersionCache[subject]; !ok {
```

Comment:
The schema client (for standalone plugins) and the schema registry (for built-in connectors) should provide caching, or no?

Reply:
Sure in Conduit we have that cached client, but in standalone plugins we will always do a gRPC call to Conduit. Granted, it's a call on the same machine, but given that this is in the hot path (executed for every record) I'm afraid it would have a noticeable effect. I remember we got quite a nice performance boost when we went from regular gRPC calls to streams for records (a long time ago), doing a call to Conduit here would essentially undo that performance gain.

Reply:
Right, that's the Conduit client. I agree, we definitely need caching.

Not for this PR, but maybe we could add caching to the sdkschema.Get()? That would benefit connectors that call the function directly too.

Reply:
Hmm, yeah, maybe we can move it. And if we move it, we can make it so it's only applied in standalone plugins, since builtin ones will already benefit from the internal cache in Conduit 👍